base/random/generation.jl

```diff
@@ -186,232 +177,169 @@ maxmultiplemix(k::UInt64) = k >> 32 != 0 ?
 end
 
-function rand(rng::AbstractRNG, g::RangeGeneratorInt{T,U}) where {T<:Integer,U<:Unsigned}
+function rand(rng::AbstractRNG, st::StateRangeInt{T,U}) where {T<:Integer,U<:Unsigned}
     x = rand(rng, U)
-    while x > g.u
+    while x > st.u
         x = rand(rng, U)
     end
-    (unsigned(g.a) + rem_knuth(x, g.k)) % T
+    (unsigned(st.a) + rem_knuth(x, st.k)) % T
+end
+
+### BigInt
+
+struct StateBigInt <: State
+    a::BigInt         # first
+    m::BigInt         # range length - 1
+    nlimbs::Int       # number of limbs in generated BigInt's (z ∈ [0, m])
+    nlimbsmax::Int    # max number of limbs for z+a
+    mask::Limb        # applied to the highest limb
 end
 
-function rand(rng::AbstractRNG, g::RangeGeneratorBigInt)
-    x = MPZ.realloc2(g.nlimbsmax*8*sizeof(Limb))
-    limbs = unsafe_wrap(Array, x.d, g.nlimbs)
+function State(::AbstractRNG, r::UnitRange{BigInt})
+    m = last(r) - first(r)
+    m < 0 && throw(ArgumentError("range must be non-empty"))
+    nd = ndigits(m, 2)
+    nlimbs, highbits = divrem(nd, 8*sizeof(Limb))
+    highbits > 0 && (nlimbs += 1)
+    mask = highbits == 0 ? ~zero(Limb) : one(Limb)<<highbits - one(Limb)
+    nlimbsmax = max(nlimbs, abs(last(r).size), abs(first(r).size))
+    return StateBigInt(first(r), m, nlimbs, nlimbsmax, mask)
+end
+
+function rand(rng::AbstractRNG, st::StateBigInt)
+    x = MPZ.realloc2(st.nlimbsmax*8*sizeof(Limb))
+    limbs = unsafe_wrap(Array, x.d, st.nlimbs)
     while true
         rand!(rng, limbs)
-        @inbounds limbs[end] &= g.mask
-        MPZ.mpn_cmp(x, g.m, g.nlimbs) <= 0 && break
+        @inbounds limbs[end] &= st.mask
+        MPZ.mpn_cmp(x, st.m, st.nlimbs) <= 0 && break
     end
     # adjust x.size (normally done by mpz_limbs_finish, in GMP version >= 6)
-    x.size = g.nlimbs
+    x.size = st.nlimbs
     while x.size > 0
         @inbounds limbs[x.size] != 0 && break
         x.size -= 1
     end
-    MPZ.add!(x, g.a)
-end
-
-#### arrays
-
-function rand!(rng::AbstractRNG, A::AbstractArray, g::RangeGenerator)
-    for i in eachindex(A)
-        @inbounds A[i] = rand(rng, g)
-    end
-    return A
+    MPZ.add!(x, st.a)
 end
 
-### random values from UnitRange
-
-rand(rng::AbstractRNG, r::UnitRange{<:Integer}) = rand(rng, RangeGenerator(r))
-
-rand!(rng::AbstractRNG, A::AbstractArray, r::UnitRange{<:Integer}) =
-    rand!(rng, A, RangeGenerator(r))
-
 ## random values from AbstractArray
 
-rand(rng::AbstractRNG, r::AbstractArray) = @inbounds return r[rand(rng, 1:length(r))]
-rand(                  r::AbstractArray) = rand(GLOBAL_RNG, r)
-
-### arrays
-
-function rand!(rng::AbstractRNG, A::AbstractArray, r::AbstractArray)
-    g = RangeGenerator(1:(length(r)))
-    for i in eachindex(A)
-        @inbounds A[i] = r[rand(rng, g)]
-    end
-    return A
-end
-
-rand!(A::AbstractArray, r::AbstractArray) = rand!(GLOBAL_RNG, A, r)
+State(rng::AbstractRNG, r::AbstractArray) = StateSimple(r, State(rng, 1:length(r)))
```

Comment:
@mschauer The performace hit happens between here and below in `rand(rng::AbstractRNG, st::StateSimple{<:AbstractArray,<:State})`. `StateSimple` introduces a kind of indirection compared to how this was implemented before (in a call like `rand(rng, [1, 2, 3])`), which currently is not eliminated/optimized-out (but could be in theory, I think).

Reply:
it reminded me of `https://github.com/JuliaGraphs/LightGraphs.jl/blob/master/src/graphtypes/simplegraphs/simpleedgeiter.jl#L85` where the wrapper `SimpleEdgeIter` is completely optimized away, but only if `has_edge` is *not* inlined.

Reply:
Oh... I don't understand how this works, but will then try to put some `@noinline` here or there!

Reply:
Cargo cult programming :-D